codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -110,49 +115,81 @@ public ByteBuf cumulate(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf in)
     public static final Cumulator COMPOSITE_CUMULATOR = new Cumulator() {
         @Override
         public ByteBuf cumulate(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf in) {
-            ByteBuf buffer;
+            CompositeByteBuf composite = null;
             try {
-                if (cumulation.refCnt() > 1) {
-                    // Expand cumulation (by replace it) when the refCnt is greater then 1 which may happen when the
-                    // user use slice().retain() or duplicate().retain().
-                    //
-                    // See:
-                    // - https://github.com/netty/netty/issues/2327
-                    // - https://github.com/netty/netty/issues/1764
-                    buffer = expandCumulation(alloc, cumulation, in.readableBytes());
-                    buffer.writeBytes(in);
-                } else {
-                    CompositeByteBuf composite;
-                    if (cumulation instanceof CompositeByteBuf) {
-                        composite = (CompositeByteBuf) cumulation;
-                    } else {
-                        composite = alloc.compositeBuffer(Integer.MAX_VALUE);
-                        composite.addComponent(true, cumulation);
+                if (cumulation instanceof CompositeByteBuf && cumulation.refCnt() == 1) {
+                    composite = (CompositeByteBuf) cumulation;
+                    if (composite.readerIndex() >= MIN_BYTES_TO_DISCARD) {
+                        composite.discardReadComponents();
+                    }
+                    // Writer index must equal capacity if we are going to "write"
+                    // new components to the end
+                    if (composite.writerIndex() != composite.capacity()) {
+                        composite.capacity(composite.writerIndex());
                     }
-                    composite.addComponent(true, in);
-                    in = null;
-                    buffer = composite;
+                } else {
+                    composite = alloc.compositeBuffer(Integer.MAX_VALUE);
+                    addFlattenedComponents(composite, cumulation);
                 }
-                return buffer;
+                addFlattenedComponents(composite, in);
+                in = null;
+                return composite;
             } finally {
                 if (in != null) {
-                    // We must release if the ownership was not transferred as otherwise it may produce a leak if
-                    // writeBytes(...) throw for whatever release (for example because of OutOfMemoryError).
+                    // We must release if the ownership was not transferred as otherwise it may produce a leak
                     in.release();
+                    if (composite != null && composite != cumulation) {
```

Comment:
should this be moved out of the `in != null` block ?

Reply:
@normanmaurer this was actually intentional since I think we only want to release it in the case we allocated it within the method _and_ aren't returning successfully.  Maybe it would be more readable to move it out and add `composite = null;` next to the existing `in = null;` ... I'll change that.

Reply:
Oops no that won't work of course since we need to return `composite` :) I will just add a comment, hopefully that will be sufficient.